### PR TITLE
feat: Make cacert export save only CA certificates

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Fixes `cacert:export` command to save only CA certificate from the certificate chain of the `self-signed-certificate` secret.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16773

### CODE REVIEW

File: src/api/che.ts [hunk 1]
```diff
@@ -120,6 +120,20 @@ export class CheHelper {
    * If secret doesn't exist, undefined is returned.
    */
   async retrieveCheCaCert(cheNamespace: string): Promise<string | undefined> {
+    const cheCaSecretContent = await this.getCheSelfSignedSecretContent(cheNamespace)
+    if (!cheCaSecretContent) {
+      return
+    }
+
+    const rootCertStartPos = cheCaSecretContent.lastIndexOf('-----BEGIN CERTIFICATE-----')
```
Comment (AndrienkoAleksandr) [inline src/api/che.ts:128]: Are you sure, that this last one certificate will be ca?

Comment (mmorhun) [reply src/api/che.ts:128]: If the chain has at leatt two certs - then yes

File: src/commands/cacert/export.ts [hunk 2]
```diff
@@ -36,22 +34,23 @@ export default class Export extends Command {
       env: 'CHE_CA_CERT_LOCATION',
       default: ''
     }),
+    [ACCESS_TOKEN_KEY]: accessToken,
```
Comment (tolusha) [inline src/commands/cacert/export.ts:37]: do we really need to provide access token?

Comment (mmorhun) [reply src/commands/cacert/export.ts:37]: Yes, as we need to read the secret from Che namespace

Comment (tolusha) [reply src/commands/cacert/export.ts:37]: I don't understand why ^(

Comment (mmorhun) [reply src/commands/cacert/export.ts:37]: Ah, yes, you are right. Just rechecked, it works without it. Thank you, will delete it.

**Comment by mmorhun** [comment]
Rebased

File: yarn.lock [hunk 3]
```diff
@@ -294,7 +294,6 @@
 "@kubernetes/client-node@0.11.2":
   version "0.11.2"
   resolved "https://registry.yarnpkg.com/@kubernetes/client-node/-/client-node-0.11.2.tgz#5dbc1d66d3c954af0f3bd294dece884737f5b761"
-  integrity sha512-Uhwd2y2qCvugICnHRC5h2MT5vw0a1dJPVVltVwmkeMuyGTPBccsTtpTcSfSLitwOrh4yr+9wG5bRcMdgeRjYPw==
```
Comment (tolusha) [inline yarn.lock:297]: I don't know why but we have a lot of removed lines like this

File: src/commands/cacert/export.ts [hunk 4]
```diff
@@ -41,23 +39,23 @@ export default class Export extends Command {
...
-    const apiTasks = new ApiTasks()
-    const tasks = new Listr([], { renderer: 'silent' })
 
-    tasks.add(apiTasks.testApiTasks(flags, this))
-    tasks.add(cheTasks.verifyCheNamespaceExistsTask(flags, this))
+    if (!await kube.hasReadPermissionsForNamespace(flags.chenamespace)) {
+      throw new Error(`E_PERM_DENIED - Permission for Che server namespace "${flags.chenamespace}" are required`)
+    }
+    if (!await cheHelper.cheNamespaceExist(flags.chenamespace)) {
+      throw new Error(`E_BAD_NS - Namespace ${flags.chenamespace} does not exist. Please specify it with --chenamespace flag`)
```
Comment (tolusha) [inline src/commands/cacert/export.ts:49]: Could you fix the error here
https://github.com/che-incubator/chectl/blob/master/src/tasks/che.ts#L496
`The configuration cannot ..` should be removed.

File: yarn.lock (hunk 3 above)
Comment (mmorhun) [reply yarn.lock:297]: Strange, I just use yarn as usual.... Will revert these changes

File: src/commands/cacert/export.ts [hunk 5]
```diff
@@ -41,23 +39,23 @@ export default class Export extends Command {
...
+    const kube = new KubeHelper(flags)
     const cheHelper = new CheHelper(flags)
-    const cheTasks = new CheTasks(flags)
-    const apiTasks = new ApiTasks()
-    const tasks = new Listr([], { renderer: 'silent' })
 
-    tasks.add(apiTasks.testApiTasks(flags, this))
-    tasks.add(cheTasks.verifyCheNamespaceExistsTask(flags, this))
+    if (!await kube.hasReadPermissionsForNamespace(flags.chenamespace)) {
+      throw new Error(`E_PERM_DENIED - Permission for Che server namespace "${flags.chenamespace}" are required`)
```
Comment (tolusha) [inline src/commands/cacert/export.ts:46]: `E_PERM_DENIED - Permission denied: no read access to '${flags.chenamespace}' namespace`

Comment (mmorhun) [reply src/commands/cacert/export.ts:46]: ok

**Review by tolusha — APPROVED** [review]

File: yarn.lock (hunk 3 above)
Comment (mmorhun) [reply yarn.lock]: From what I see, yarn deletes them automatically.

**Review by flacatus — APPROVED** [review]

**Review by AndrienkoAleksandr — APPROVED** [review]

File: yarn.lock (hunk 3 above)
Comment (AndrienkoAleksandr) [reply yarn.lock]: > integrity sha512-Uhwd2y2qCvugICnHRC5h2MT5vw0a1dJPVVltVwmkeMuyGTPBccsTtpTcSfSLitwOrh4yr+9wG5bRcMdgeRjYPw==

This thing generation depends on yarn version

**Comment by benoitf** [comment]
hello, which version of yarn do you have ?
it's removing all `integrity sha512-xxx` lines in yarn.lock ?

File: src/api/che.ts [hunk 6]
```diff
@@ -120,6 +121,35 @@ export class CheHelper {
...
+    const pemEndHeader = '-----END CERTIFICATE-----'
+    const certRegExp = new RegExp(`(^${pemBeginHeader}$(?:(?!${pemBeginHeader}).)*^${pemEndHeader}$)`, 'mgs')
+    const certsPem = cheCaSecretContent.match(certRegExp)
+
+    const caCertsPem: string[] = []
+    if (certsPem) {
+      for (const certPem of certsPem) {
+        const cert = nodeforge.pki.certificateFromPem(certPem)
+        const basicConstraintsExt = cert.getExtension('basicConstraints')
+        if (basicConstraintsExt && (basicConstraintsExt as any).cA) {
```
Comment (tolusha) [inline src/api/che.ts:139]: 'cA` is it correct?

Comment (mmorhun) [reply src/api/che.ts:139]: Sure. Otherwise it wouldn't work. (I also wondered why they named the field such way).

**Comment by mmorhun** [comment]
@benoitf I have 1.6.0 which I must update.

**Review by AndrienkoAleksandr — APPROVED** [review]

**Pull request merged**